pkg/git/libgit2/checkout.go

```diff
@@ -69,97 +69,148 @@ type CheckoutBranch struct {
-	defer upstreamCommit.Free()
+		upstreamCommit, err := repo.LookupCommit(branch.Target())
+		if err != nil {
+			return nil, fmt.Errorf("unable to lookup commit '%s' for '%s': %w",
+				c.Branch, managed.EffectiveURL(url), gitutil.LibGit2Error(err))
+		}
+		defer upstreamCommit.Free()
 
-	// Once the index has been updated with Fetch, and we know the tip commit,
-	// a hard reset can be used to align the local worktree with the remote branch's.
-	err = repo.ResetToCommit(upstreamCommit, git2go.ResetHard, &git2go.CheckoutOptions{
-		Strategy: git2go.CheckoutForce,
-	})
-	if err != nil {
-		return nil, fmt.Errorf("unable to hard reset to commit for '%s': %w", managed.EffectiveURL(url), gitutil.LibGit2Error(err))
-	}
+		// Once the index has been updated with Fetch, and we know the tip commit,
+		// a hard reset can be used to align the local worktree with the remote branch's.
+		err = repo.ResetToCommit(upstreamCommit, git2go.ResetHard, &git2go.CheckoutOptions{
+			Strategy: git2go.CheckoutForce,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("unable to hard reset to commit for '%s': %w", managed.EffectiveURL(url), gitutil.LibGit2Error(err))
+		}
 
-	// Use the current worktree's head as reference for the commit to be returned.
-	head, err := repo.Head()
-	if err != nil {
-		return nil, fmt.Errorf("git resolve HEAD error: %w", err)
-	}
-	defer head.Free()
+		// Use the current worktree's head as reference for the commit to be returned.
+		head, err := repo.Head()
+		if err != nil {
+			return nil, fmt.Errorf("git resolve HEAD error: %w", err)
+		}
+		defer head.Free()
 
-	cc, err := repo.LookupCommit(head.Target())
-	if err != nil {
-		return nil, fmt.Errorf("failed to lookup HEAD commit '%s' for branch '%s': %w", head.Target(), c.Branch, err)
+		cc, err := repo.LookupCommit(head.Target())
+		if err != nil {
+			return nil, fmt.Errorf("failed to lookup HEAD commit '%s' for branch '%s': %w", head.Target(), c.Branch, err)
+		}
+		defer cc.Free()
+
+		return buildCommit(cc, "refs/heads/"+c.Branch), nil
+	} else {
+		repo, err := git2go.Clone(url, path, &git2go.CloneOptions{
+			FetchOptions: git2go.FetchOptions{
+				DownloadTags:    git2go.DownloadTagsNone,
+				RemoteCallbacks: RemoteCallbacks(ctx, opts),
+				ProxyOptions:    git2go.ProxyOptions{Type: git2go.ProxyTypeAuto},
+			},
+			CheckoutOptions: git2go.CheckoutOptions{
+				Strategy: git2go.CheckoutForce,
+			},
+			CheckoutBranch: c.Branch,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("unable to clone '%s': %w", managed.EffectiveURL(url), gitutil.LibGit2Error(err))
+		}
+		defer repo.Free()
+		head, err := repo.Head()
+		if err != nil {
+			return nil, fmt.Errorf("git resolve HEAD error: %w", err)
+		}
+		defer head.Free()
+		cc, err := repo.LookupCommit(head.Target())
+		if err != nil {
+			return nil, fmt.Errorf("failed to lookup HEAD commit '%s' for branch '%s': %w", head.Target(), c.Branch, err)
+		}
+		defer cc.Free()
+		return buildCommit(cc, "refs/heads/"+c.Branch), nil
```

Comment:
Since these changes bring back the old checkout implementation along with the new one, it'd be better to put then into separate functions which can then be tested separately to verify that both of them work.